### PR TITLE
Switching to rollup and adding errorBoundary opt

### LINKED ISSUE #74: Central handling of the uncaught application errors
Currently single-spa doesn't care much about errors that are happening within applications after completion of the "mount" hook. In our [Isomorphic Layout Composer](https://github.com/namecheap/ilc) project we solved this by passing [error handler function](https://github.com/namecheap/ilc/blob/master/ilc/client.js#L62) to the app via props which will be handled in adapter and receive all application specific errors that were not handled (see [example here](https://github.com/namecheap/ilc/blob/master/adapters/vuejs/src/single-spa-vue.js#L28)).

While this is not an issue for us - it would be great to see such functionality incorporated into adapter interface as "errorHandler" function.

### PR DESCRIPTION
**Pull request by joeldenning**
Resolves #74. In all of this I ended up switching the build from babel cli to rollup + terser. The bundle size has gone down from 11kb to 5kb, but there are some weird things about it (see below)

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -1,14 +1,4 @@
 {
   "presets": ["@babel/preset-env", "@babel/preset-react"],
-  "plugins": [
-    [
-      "@babel/plugin-transform-modules-umd",
```
Comment (joeldenning) [inline .babelrc:5]: handled by rollup now

File: package.json [hunk 2]
```diff
@@ -51,7 +56,10 @@
     "pretty-quick": "^2.0.1",
     "react": "^16.13.1",
     "react-dom": "^16.13.1",
-    "rimraf": "^3.0.2"
+    "rollup": "^2.6.1",
+    "rollup-plugin-babel": "^4.4.0",
+    "rollup-plugin-terser": "^5.3.0",
+    "single-spa-react": "."
```
Comment (joeldenning) [inline package.json:62]: This is to trick webpack into accepting a self-referential import. Ideally we'd do this with the package.json exports field, but it's not yet supported in react. Below is the explanation

We want this:
```js
!(function (e, t) {
  "object" == typeof exports && "undefined" != typeof module
    ? (module.exports = t(require("react"), require("../lib/single-spa-react.js")))
    : "function" == typeof define && define.amd
    ? define(["react", "single-spa-react"], t)
    : ((e = e || self).Parcel = t(e.React, e.singleSpaReact));
})(this, function (e, t) {
  "use strict";
```

But rollup can only give us this (afaik):

```js
!(function (e, t) {
  "object" == typeof exports && "undefined" != typeof module
    ? (module.exports = t(require("react"), require("single-spa-react")))
    : "function" == typeof define && define.amd
    ? define(["react", "single-spa-react"], t)
    : ((e = e || self).Parcel = t(e.React, e.singleSpaReact));
})(this, function (e, t) {
  "use strict";
```

The second one works if `require('single-spa-react')` works as a self referential import. Note that it causes the following yarn warning, but did work for me with yarn install + yarn link:

```
warning Using the "file:" protocol implicitly is deprecated. Please either prepend the protocol or prepend the path "." with "./".
```

File: src/single-spa-react.test.js [hunk 3]
```diff
@@ -372,6 +373,63 @@ describe("single-spa-react", () => {
...
+
+      return lifecycles
+        .bootstrap()
+        .then(() => lifecycles.mount(props))
+        .then(() => {
+          return expect(console.warn).not.toHaveBeenCalled();
+        });
+    });
+
+    it(`should call opts.errorBoundary during an error boundary handler`, () => {
```
Comment (joeldenning) [inline src/single-spa-react.test.js:395]: This test has me feeling queasy - we should consider switching to `@testing-library/react`

**Comment by joeldenning** [comment]
Another thing to note is that I verified that global UMD usage of both singleSpaReact and Parcel are unchanged. You can see the code sandbox at https://codesandbox.io/s/damp-platform-5kn03?file=/app.js

File: package.json (hunk 2 above)
Comment (joeldenning) [reply package.json:62]: I tried the following rollup config which I thought would work, but it didn't:

```
  {
    input: "src/parcel.js",
    output: {
      file: "parcel/index.js",
      format: "umd",
      name: "Parcel",
      sourcemap: true,
      globals: {
        react: "React",
        "single-spa-react": "singleSpaReact",
      },
    },
    plugins: [babel(), terser()],
    external: {
      "single-spa-react": "../lib/single-spa-react",
      React: "react",
    },
  },
```

File: src/single-spa-react.js [hunk 4]
```diff
@@ -248,3 +239,72 @@ function reactDomRender({ opts, elementToRender, domElement, whenFinished }) {
...
+    );
+  }
+
+  return elementToRender;
+}
+
+function createErrorBoundary(opts) {
+  // Avoiding babel output for class syntax and super()
+  // to avoid bloat
+  function SingleSpaReactErrorBoundary(props) {
```
Comment (TheMcMurder) [inline src/single-spa-react.js:271]: This is interesting. Just to make sure I’m following, here you’re creating a JS class (which is actually a function) directly with a function to reduce bundle size?

What (if anything) does the normal ‘...extends react.component’ part of the class declaration give us that this is missing?

Comment (joeldenning) [reply src/single-spa-react.js:271]: The behavior is the same as using class syntax afaik. I'm simulating the super() call below

Comment (TheMcMurder) [reply src/single-spa-react.js:273]: Very fancy.

**Pull request merged**